packages/react-router/README.md

```diff
@@ -0,0 +1,378 @@
+
+```tsx
+import {useEffect} from 'react';
+import {useCurrentUrl} from '@shopify/react-router';
+
+function useNavigationTracking() {
+  const currentUrl = useCurrentUrl();
+
+  useEffect(() => {
+    trackButNotInACreepyWayPlease(currentUrl.href);
+  }, [currentUrl]);
+}
+```
+
+You may be tempted to use this hook in components nested under a `Route` to get details about the URL, such as the path or querystring. While this use case is completely valid, it can make the application feel overly coupled to the URL in ways that are hard to understand. An alternative approach extracts those details just once, in the `Route`’s `render` prop, passes it in to the "top level" component, which can then distribute those details to its children through props or context:
+
+```tsx
+import {Route} from '@shopify/react-router';
+
+function MyComponent({create = false, prefillTitle = ''}) {
+  // Use the props, just like you do for any other prop!
+}
+
+function MyComponentRoute() {
+  return (
+    <Route
+      match={/^\/resource(\/new)?/}
+      render={({pathname, searchParams}) => (
+        <MyComponent
+          create={pathname.contains('/new')}
+          prefillTitle={searchParams.get('prefillTitle')}
+        />
+      )}
+    />
+  );
+}
+```
+
+#### `EnhancedURL`
+
+This library is driven by a the `URL` object. In order to support features like location state, however, most of this library operates on `EnhancedURL` objects. These objects are identical to a `URL` object except for the following additions:
+
+- All members should be considered immutable
+- An additional `state` field is provided, which is the location state of the location the object represents
+
+### `useRouter()`
+
+Not all navigation happens by users clicking on links. Applications often need to programmatically navigate in response to things like resources being created or updated. `@shopify/react-router` provides the `useRouter` hook for this purpose. The hook provides the singleton router instance.
+
+```tsx
+import React from 'react';
+import {useRouter} from '@shopify/react-router';
+
+function MyComponent() {
+  const router = useRouter();
+  return (
+    <CreateProduct onComplete={(id) => router.navigate(`/products/${id}`)} />
+  );
+}
+```
+
+The router instance has the following methods:
+
+#### `router.navigate()`
+
+`router.navigate()` is the main method of the router. It allows you to add or replace an entry in the navigation stack. The first argument is a description of the location to push to, and can be one of the following:
+
+- A string, which should be a relative or absolute pathname, search, and hash,
+- A `URL` object (**note:** must be on the same origin as the current URL),
```

Comment:
Why can't the router be smart enough to fall back to setting `location` when it encounters a different base URL? The behaviour could even be configurable. I've always found it a wart of RR that you have to remember when to use it vs `window.location`, especially in server rendered apps where you are generally avoiding `window.location` in other places.

Reply:
Good idea 👍